### DISCUSSION TITLE
ImageMagick throws a warning resulting in slow conversion of a 200x200 dpi tiff to 300x300 dpi tiff.

### DISCUSSION THREAD

**Discussion by pprasanthk** [ImageMagick/ImageMagick#3946] [Help]
### ImageMagick version

7.0.11

### Operating system

Windows

### Operating system, version and so on

Any version of Windows OS.

### Description

Hi, 

We are using ImageMagick 7.0.11 version. Attached is the zip file which contains our input file **test.tif**, which is of 200x200 dpi resolution. 

We are trying to convert this **test.tif** file to a black and white TIFF fil with 300x300 dpi resolution. When we are doing that we are getting the below warning from ImageMagick library at the call to image.Read(<input filename>);  Due to this, conversion is taking a lot of time.

**warning: Magick: Unknown field with tag 33000 (0x80e8) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/1036**

Below is the code snippet to reproduce this error:
**Image image; // class MagickPPExport Image declared inside Magick++/Image.h
try {
	// Read a file into Magick image object
	image.read( sInputSinglePage.c_str() );
}
catch( Warning &error_ ) {
	cout << "Caught warning: " << error_.what() << endl;
}**

Could you please let us know how can we get rid of this warning to increase the above mentioned file conversion?

Please let us know if you need any more information.

[test.zip](https://github.com/ImageMagick/ImageMagick/files/6829748/test.zip)

--
Thanks
Prasanth

### Steps to Reproduce

**Image image; // class MagickPPExport Image declared inside Magick++/Image.h
try {
	// Read a file into Magick image object
	image.read( sInputSinglePage.c_str() );
}
catch( Warning &error_ ) {
	cout << "Caught warning: " << error_.what() << endl;
}**

### Images

[test.zip](https://github.com/ImageMagick/ImageMagick/files/6829854/test.zip)

**Comment by dlemstra** [comment]
You can do `image.quiet()` before calling `.read()` to silence a warning.

**Reply by pprasanthk** [reply]
Hi, 

We tried to use image.quiet(), but it did not increase the conversion speed. Here are the few observations from our side: 

[test.zip](https://github.com/ImageMagick/ImageMagick/files/6847666/test.zip)

We tried to convert the input TIFF image (attached here with for your reference) into another black and white TIFF image with 300x300 resolution.
1.	We tried to do the above conversion with ImageMagick version 7.0.8-67 and noticed that the conversion took around 10 seconds.
2.	When we tried to the same conversion with ImageMagick version 7.0.11, we noticed that this conversion is taking more than 60 seconds.

Could you please let us know what changed from the side of the ImageMagick library which is causing in this conversion delay? 

--
Thanks
Prasanth

**Reply by dlemstra** [reply]
Various things could have been changed. What happens when you upgrade to the most recent version?

**Comment by snibgo** [comment]
The thread title includes:
> slow conversion of a 200x200 dpi tiff to 300x300 dpi tiff

... but the code shown merely reads a file.

I doubt if the warning causes noticable performance problems.

Changing dpi (eg with "-density" at the CLI) changes metadata only. So it takes no noticable time, compared to reading and writing the pixels.

I suppose your code is doing something else. What?

Please also show the text output from "magick -version" for your two versions. Perhaps one is HDRI and the other is not.

**Comment by fmw42** [comment]
The file is compressed TIF with many pages. ImageMagick will decompress the file (all the pages), then resize, then write the file. But the default for TIFF is to write uncompressed, if no compression is provided. This will mean writing the file will take longer and be a larger file. Perhaps this is an issue?

**Reply by pprasanthk** [reply]
We are using ImageMagick APIs to convert each page once to a B&W 300x300 TIFF. 

As you asked previously, I have tested with the latest version of ImageMagick (7.1.0.4) and noticed that this conversion is taking around half a minute, while version ImageMagick 7.0.11 is taking around a minute (double the time latest version is taking). 
Still this is a considerable amount of time, given that ImageMagick version 7.0.8-67 is taking around 10 seconds. 

Also, we do not have HDRI support included in any of the versions of ImageMagick that we are using.

Could you please let us know what slowed down the conversion time taken?

--
Thanks
Prasanth

**Comment by snibgo** [comment]
> Could you please let us know what slowed down the conversion time taken?

You can answer the question better than we can.

I suggest you insert code that displays a timestamp at each stage of your program. This will tell you where the program has become slower.

**Reply by pprasanthk** [reply]
`int TiffAppend(string sResultSinglePage, string sResult)
{
	TIFF *in, *out;
	in = TIFFOpen(sResultSinglePage.c_str(), "r");
	if (in != NULL) 
	{
		out = TIFFOpen(sResult.c_str(), "a");
		if (out == NULL)
			return (-2);
		if (!tiffcp(in, out)) // LibTiff sample code Tiff copy
			return (-1);
		TIFFClose(out);
		(void) TIFFClose(in);
		remove(sResultSinglePage.c_str());
	}
	return (0);
}

bool ApplyResolutionConversionRules(const int dx, const int dy, const int sx, const 
int sy, const char* const gNewSizeText = NULL, 
const Point* const pNewDensity = NULL, Image* const image = NULL)
{
bool isConversionRequired = false;
	if (g_nResetResolution == RESOLUTION_CHANGE_WITH_RESIZE)
	isConversionRequired = true;
	if(image == NULL)
	{
		return isConversionRequired;
	}
	// Resize to requested resolution (density)
	image->resize(gNewSizeText); 
	image->density(*pNewDensity);
	image->quality(g_nJpegQuality);
	image->resolutionUnits(PixelsPerInchResolution);
	return isConversionRequired;
}
void ConvertSinglePage (string sInputSinglePage, string sResult, int nPageCount)
{
	Image image;
	int nColorModePage = 2;
	string sColorModePage = "Color";	// For log-file only
	string sResultSinglePage = sResult + POSTFIX_SINGLEPAGE_RESULT;

	try 
	{
		try {
			image.quiet();
			image.read( sInputSinglePage.c_str() );
		}
		catch( Warning &error_ )
		{
			cout << "Caught warning: " << error_.what() << endl;
		}

	    float requested_x = 0, requested_y = 0;
		image.autoOrient();		
		Point gImgDensity = image.density();
		int dx = gImgDensity.x();
		int dy = gImgDensity.y();

		Geometry gImgSize = image.size();
		int sx = gImgSize.width();
		int sy = gImgSize.height();

		// Get the color mode of this page
		if (image.type() == BilevelType)
		{
			nColorModePage = 0;
			sColorModePage = "B/W";
		}
		else if (image.type() == GrayscaleType || image.type() == GrayscaleAlphaType)
		{
			nColorModePage = 1;	// Grey
			sColorModePage = "Grey";
		}
		else 
		{
			nColorModePage = 2;	// Color, undefined, everything else
			sColorModePage = "Color";
		}
		char gNewSizeText[52];
		sprintf_s(gNewSizeText, 50, "%dx%d!", g_nReqResolution*sx/dx, g_nReqResolution*sy/dy);
		Point gNewDensity(g_nReqResolution, g_nReqResolution);
	ApplyResolutionConversionRules(dx,dy,sx,sy,gNewSizeText,&gNewDensity, &image);
	
		image.magick("TIFF");
		if ( g_sReqColor=="2" && 			// Requested: Color
		( g_nColorUpsample ||	// either always (including upsample)
		(!g_nColorUpsample && nColorModePage==2) ) )// if page already color
		{
			image.compressType(JPEGCompression);
			image.type(TrueColorType);
			image.depth(8);
		}
		else if ((g_sReqColor=="1" || g_sReqColor=="2") &&// Grey or Color
		( g_nColorUpsample || // either always (including upsample)
		(!g_nColorUpsample && nColorModePage>=1) )) // or only if page already grey or color
		{
			image.compressType(JPEGCompression);
			image.type(GrayscaleType);
			image.depth(8);
		}
		else
		{
			image.compressType(Group4Compression);
			image.type(BilevelType);
		}
		if(scaleto> -1 && scaleto <= 20)
		{
			requested_x=SCALETO_CONST[scaleto].x;
			requested_y=SCALETO_CONST[scaleto].y;
			sprintf_s(gNewSizeText, 50, "%dx%d!",(int)(g_nReqResolution*requested_x),(int)(g_nReqResolution*requested_y));
			image.resize(gNewSizeText);
		}
		else if(scaleto != -1 && scaleto > 23)
		{
			requested_x = SCALETO_CONST[A4].x;
			requested_y = SCALETO_CONST[A4].y;
			sprintf_s(gNewSizeText, 50, "%dx%d!", (int)(g_nReqResolution*requested_x), (int)(g_nReqResolution*requested_y));
			image.resize(gNewSizeText);
		}
		// Result
		gImgDensity = image.density();
		dx = gImgDensity.x();
		dy = gImgDensity.y();

		gImgSize = image.size();
		sx = gImgSize.width();
		sy = gImgSize.height();

		if (nPageCount > 1)
		{
			// Multiple pages
			image.write(sResultSinglePage.c_str());
			TiffAppend(sResultSinglePage, sResult);
		}
		else
		{
			// Single page
			image.write(sResult.c_str());
		}
	}
	catch( Exception &error_ )
	{
		cout << "Caught exception: " << error_.what() << endl;
		g_nResult = 1;
	}
}
`

**Reply by pprasanthk** [reply]
Below calls in the above piece of code are increasing the conversion time: 
1. image.type(BilevelType);
2. image.write(sResultSinglePage.c_str()); // inside if (nPageCount > 1) { }
3. image->resize(gNewSizeText); // inside ApplyResolutionConversionRules() method

**Comment by fmw42** [comment]
I did some time tests on Mac OSX Sierra using the current ImageMagick 7.1.0.4 and ImageMagick 7.0.8.67 and find the current version runs much faster (0.7 s) than the older version (4 s). Here are my times (run 3 times each)

```
time im7 magick -quiet test.tif -compress group4 test2.tif

real	0m0.783s
user	0m0.482s
sys	0m0.278s

time im7 magick -quiet test.tif -compress group4 test2.tif

real	0m0.727s
user	0m0.484s
sys	0m0.225s

time im7 magick -quiet test.tif -compress group4 test2.tif

real	0m0.735s
user	0m0.487s
sys	0m0.232s




time im70867 magick -quiet test.tif -compress group4 test2.tif

real	0m3.968s
user	0m12.898s
sys	0m1.002s

time im70867 magick -quiet test.tif -compress group4 test2.tif

real	0m3.978s
user	0m12.396s
sys	0m1.196s

time im70867 magick -quiet test.tif -compress group4 test2.tif

real	0m3.983s
user	0m12.712s
sys	0m1.094s

```

**Comment by snibgo** [comment]
I also get a performance increase in v7.1.0-4 compared to 7.0.10-45, and in 7.0.10-45 compared to 7.0.8-64.

@pprasanthk: When pasting code, please insert a line with three backticks above and below the code. Then the formatting is not disturbed.

As I suspected, you are not merely changing the resolution, but you are also resizing the image.

In your OP, you said the problem was image.read(). But now you say it is type(), write() and resize(). How much have these changed? Please show actual timings.

Please also show the text output from "magick -version" for the versions you use.

One overall difference may be the use of OpenCL.

There have been many changes over the versions you mention. For example, quantize.c, which is used by image.type(BilevelType).

**Comment by fmw42** [comment]
Please show actual code used, not just your read! If possible, test in command line as well to see if it is your code or if you get similar timings. Also check your resources to see that there is no bottleneck.

Snibgo's point about OpenCL is good. Also sometimes OpenMP can cause issues depending upon architecture.

**Reply by pprasanthk** [reply]
[src.zip](https://github.com/ImageMagick/ImageMagick/files/6863872/src.zip)
Hi, 

Please find the source code attached here with for your reference. Could you please let us know if I am missing  something which could speed up the conversion process?

Note: 
1. We have generated ImageMagick binaries from configure.exe, where we have not included OpenMP support.
2. main() method is there in ImageToTiff.cpp. 
3. Other files, TiffSplit.cpp and tiffsplit.c contains methods from LibTiff.
4. Command line arguments that we use are mentioned @ line# 444 inside the method int _tmain(int argc, _TCHAR* argv[]).

Please let me know if you need any more information.

--
Thanks
Prasanth
